### PR TITLE
Feat/support php8

### PR DESCRIPTION
**Pull request by Shoormax**

### CODE REVIEW

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php [hunk 1]
```diff
@@ -130,17 +122,15 @@ public function getKeys()
      *
      * @return Filter
      */
-    public function current()
+    public function current(): mixed
```
Comment (burahimu) [inline src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php:125]: **suggestion (non-blocking):**

```suggestion
    public function current(): Filter
```

🤷

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/AbstractType.php [hunk 2]
```diff
@@ -55,11 +57,7 @@ final public function getStatus($config)
      */
     final public function isExclude($config)
     {
-        if (strpos($config, '!') !== false) {
-            return true;
-        }
-
-        return false;
+        return str_contains($config, '!');
```
Comment (burahimu) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/AbstractType.php:60]: **praise:** 💯 🥇

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/AbstractType.php [hunk 3]
```diff
@@ -55,11 +57,7 @@ final public function getStatus($config)
      */
     final public function isExclude($config)
```
Comment (burahimu) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/AbstractType.php:58]:
```suggestion
    final public function isExclude($config): bool
```

🤷

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php [hunk 4]
```diff
@@ -26,7 +28,7 @@ protected function getPattern()
      */
     protected function transform($config)
     {
-        $refStatus = substr($config, 0, strlen((int) $config - 2));
+        $refStatus = substr($config, 0, strlen((string) ((int) $config - 2)));
```
Comment (burahimu) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php:31]: **question (non-blocking):** why did you need to do this? 🤔

**Review by burahimu — DISMISSED** [review]

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php (hunk 4 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php:31]: I changed the version of `m6web/php-cs-fixer-config` and it now adds this line `declare(strict_types=1);` at the beginning of each file. Because of this, an error was raised on this line because `strlen` expects its first argument to be a string.
I have to admit that I don't really understand what this line does and I didn't want to break anything so I went with this 🤷

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php [hunk 5]
```diff
@@ -26,7 +28,7 @@ protected function getPattern()
      */
     protected function transform($config)
```
Comment (fferriere) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php:29]: suggestion (non-blocking): we can add type-hints on parameters and on return type ? (anywhere)
```suggestion
    protected function transform(string $config): array
```

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php (hunk 1 above)
Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php:125]: it seems that the `$this->get(...)` function is able to return `null`

so this would be more accurate:
```suggestion
    public function current(): ?Filter
```

this could happen when the FilterCollection is empty, maybe? 🤷 
(I have no idea what this class is about...)

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php [hunk 6]
```diff
@@ -90,10 +88,8 @@ public function remove(Filter $filter)
      * Get item
      *
      * @param int $iterator
-     *
-     * @return mixed
      */
-    public function get($iterator)
+    public function get($iterator): mixed
```
Comment (Oliboy50) [inline src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php:92]:
```suggestion
    public function get($iterator): ?Filter
```


ℹ️ this can also be done for the `getByName()` function

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php (hunk 4 above)
Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php:31]: I tried it [online](https://3v4l.org/CK1RM) and it does not break the current goal of this statement 
(which seems to be counting the number of digits of an integer)

**Review by Oliboy50 — APPROVED** [review]
🤞 

=> new major release

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php (hunk 5 above)
Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php:29]: since this will be a new major release 
=> I agree that we should add all missing type-hints in this PR (in order to make a new major release for each type-hints added)

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php (hunk 1 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterCollection.php:125]: Yep, it returns `null` if the collection is empty, I changed the code accordingly

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php (hunk 5 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Matcher/Status/Type/CompleteType.php]: It's done :)

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php [hunk 7]
```diff
@@ -40,130 +31,67 @@ public function __construct($name)
...
 
-    /**
-     * setLevel
-     *
-     * @param mixed $level Log level
-     *
-     * @return Filter
-     */
-    public function setLevel($level)
+    public function setLevel(?string $level): self
```
Comment (Oliboy50) [inline src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: why not `mixed` like the others?

**Review by Oliboy50 — APPROVED** [review]
🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏 🤞 🙏

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php (hunk 7 above)
Comment (fferriere) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: Maybe, because we know the $level [type](https://github.com/BedrockStreaming/LogBridgeBundle/pull/68/files#diff-999d6f0b9190c37b81242fb5cf9dacfd247f391c5487b24b1d5781c032fb2a68R20) :blush:

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php [hunk 8]
```diff
@@ -57,169 +44,99 @@ protected function createFilter($name)
...
+    public function parse(string $name, array $config): Filter
     {
-        $filter = $this->createFilter($name);
-
-        if (!array_key_exists('route', $config) || !array_key_exists('method', $config) || !array_key_exists('status', $config)) {
+        if (
+            !array_key_exists('route', $config) ||
+            !array_key_exists('method', $config) ||
+            !array_key_exists('status', $config)
+        ) {
```
Comment (fferriere) [inline src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:58]: As we use $config['route'] and others below, I suggest this non-blocking suggestion:
```suggestion
        $route = $config['route'] ?? null;
        $method = $config['method'] ?? null;
        $status = $config['status'] ?? null;
        if (
            null === $route ||
            null === $method ||
            null === $status
        ) {
```
and replace $config['route'] by $route below

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php (hunk 7 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: `setLevel` is only used in this file `M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php` and the method looks like this 
```
protected function parseLevel(Filter $filter, ?string $level): void
    {
        if (!is_string($level) && $level !== null) {
            throw new ParseException(sprintf('Unrecognized value "%s" from level parameter', $level));
        }
        // ...
    }
```

I figured "why would I allow mixed when I know exactly what's being used ?"

Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: this $level type is also something that we changed to `?string` in this PR

so my question about `mixed` is also true for the $level type and for the `getLevel()` function

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php (hunk 8 above)
Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:58]: this change would throw an exception if the keys exist with a `null` value

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php (hunk 7 above)
Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: @Shoormax ok, but the signature of `parseLevel()` is also something that we change in this PR, it currently accepts something else than a string

Capture d’écran 2022-05-23 à 10 43 25


before this PR, if we pass an `int` as $level to `parseLevel()`, it will throw a `ParseException`

after this PR, if we pass an `int` as $level to `parseLevel()`, it will throw a `TypeError`

this may not be a big deal, but it will break `parseLevel()` callers if they only `catch (ParseException)` or `catch (\Exception)` instead of `catch (\Throwable)`

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php (hunk 8 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:58]: This won't work because we're just checking that the key exist, even if it's null.
Given `$config['route'] = null` then `!array_key_exists('route', $config)` is `false` but `($config['route'] ?? null) === null` is `true`

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php (hunk 7 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:75]: Yep, you're right, nice catch there & thanks for the explanation, I'll change it back to `mixed`

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php (hunk 8 above)
Comment (fferriere) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:58]: You're right. But we check in parseRoute if $route is not null.
It was a suggestion

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php [hunk 9]
```diff
@@ -57,169 +44,99 @@ protected function createFilter($name)
...
-    /**
-     * parseRoute
-     *
-     * @param Filter $filter Filter
-     * @param mixed  $route  Route parameter value
-     *
-     * @throws ParseException
-     */
-    protected function parseRoute(Filter $filter, $route)
+    protected function parseRoute(Filter $filter, ?string $route): void
```
Comment (fferriere) [inline src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:76]: As we throw an exception if $route is null, why we allow here ?
Non-blocking suggestion:
```suggestion
    protected function parseRoute(Filter $filter, string $route): void
```

Comment (Oliboy50) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:76]: I'd say because the code is messy... but again we should not break too much things in this PR, so I would even suggest to revert to:

```suggestion
    protected function parseRoute(Filter $filter, mixed $route): void
```

for now, to avoid throwing a `TypeError` instead of a `ParseException` when the function is misused... 

=> of course this would be a good idea to have a stricter typing (to have a code more maintainable in our side and easier to understand for its users), but maybe in another PR where we focus on these breaking changes?

Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:76]: You're right, I'll change this and the others, there's no point in allowing null & throwing an exception if it is. 
I'll push a new commit with your suggestion and all it implies

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Dumper/PhpMatcherDumper.php [hunk 10]
```diff
@@ -237,19 +146,14 @@ public function getMatchFilterKey(\$route, \$method, \$status)
             }
         }
 
-        return false;
+        return '';
```
Comment (shavounet) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/Dumper/PhpMatcherDumper.php:149]: Not a big fan of using an empty string as a "not found value" (and using falsy coercion on this, later in the code). Maybe I'm missing something, but can't we use a nullable return type ?

**Review by shavounet — APPROVED** [review]

File: composer.json [hunk 11]
```diff
@@ -8,18 +8,18 @@
     "minimum-stability" : "dev",
     "prefer-stable": true,
     "require": {
-        "php": ">=7.1.3",
+        "php": ">=8.0",
```
[the remark this answers is not in the document]
Comment (Oliboy50) [reply composer.json:11]: the symfony community seems to prefer `>=` now

source: https://github.com/symfony/framework-bundle/blob/6.1/composer.json#L19

**Review by systemasis — APPROVED** [review]

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/Dumper/PhpMatcherDumper.php (hunk 10 above)
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Matcher/Dumper/PhpMatcherDumper.php]: Changed the code to return null instead of an empty string 👍

File: src/M6Web/Bundle/LogBridgeBundle/Matcher/MatcherInterface.php [hunk 12]
```diff
@@ -9,84 +11,22 @@
...
-     * match
-     *
-     * @param string $route  Route name
-     * @param string $method Method name
-     * @param int    $status Http code status
-     *
-     * @return bool
-     */
-    public function match($route, $method, $status);
+    public function match(string $route, string $method, int $status): bool;
```
Comment (shavounet) [inline src/M6Web/Bundle/LogBridgeBundle/Matcher/MatcherInterface.php:14]: For coherency, shouldn't we use mixed here too ?

File: src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php [hunk 13]
```diff
@@ -1,36 +1,27 @@
...
-    private $options;
+    private array $options;
 
-    /**
-     * __construct
-     *
-     * @param string $name name Filter name
-     */
-    public function __construct($name)
+    public function __construct(string $name)
```
[the remark this answers is not in the document]
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/Filter.php:24]: 👍

File: src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php [hunk 14]
```diff
@@ -24,31 +25,17 @@ class FilterParser
...
-    /** @var string */
-    protected $filterClass;
+    protected string $filterClass;
 
-    /**
-     * __construct
-     *
-     * @param RouterInterface $router Router service
-     */
     public function __construct(RouterInterface $router = null)
```
[the remark this answers is not in the document]
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/Config/FilterParser.php:32]: 👍

File: src/M6Web/Bundle/LogBridgeBundle/EventDispatcher/BuilderEvent.php [hunk 15]
```diff
@@ -10,68 +12,36 @@
...
-    private $config;
+    private array $config;
 
-    /**
-     * __construct
-     *
-     * @param BuilderInterface $builder Builder
-     * @param array            $config  Config
-     */
     public function __construct(BuilderInterface $builder, array $config = [])
```
[the remark this answers is not in the document]
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/EventDispatcher/BuilderEvent.php:19]: 👍

File: src/M6Web/Bundle/LogBridgeBundle/EventDispatcher/LogExceptionListener.php [hunk 16]
```diff
@@ -9,26 +11,20 @@
...
 {
-    /** @var string */
-    protected $requestExceptionAttribute;
+    protected string $requestExceptionAttribute;
 
-    /**
-     * @param string $requestExceptionAttribute
-     */
-    public function __construct($requestExceptionAttribute)
+    public function __construct(string $requestExceptionAttribute)
```
[the remark this answers is not in the document]
Comment (Shoormax) [reply src/M6Web/Bundle/LogBridgeBundle/EventDispatcher/LogExceptionListener.php:16]: 👍 I changed the code where you commented and changed some other files for the same reason, thanks for your review :)

**Review by FloFrad — APPROVED** [review]
